### ISSUE TITLE
SocketBuffedWriter flushToTarget Broken Pipe

### ISSUE THREAD

**Issue by wangshuzheng** [housepower/ClickHouse-Native-JDBC#100]
### JDBC version in pom.xml
8

### ClickHouse version
1.7-stable

### Error  logs
```
java.sql.SQLException: Broken pipe (Write failed)
	at com.github.housepower.jdbc.connect.PhysicalConnection.disPhysicalConnection(PhysicalConnection.java:103)
	at com.github.housepower.jdbc.ClickHouseConnection.serverInfo(ClickHouseConnection.java:184)
	at com.github.housepower.jdbc.ClickHouseConnection.createPhysicalInfo(ClickHouseConnection.java:164)
	at com.github.housepower.jdbc.ClickHouseConnection.createClickHouseConnection(ClickHouseConnection.java:159)
	at com.github.housepower.jdbc.NonRegisterDriver.connect(NonRegisterDriver.java:23)
	at com.alibaba.druid.pool.DruidAbstractDataSource.createPhysicalConnection(DruidAbstractDataSource.java:1643)
	at com.alibaba.druid.pool.DruidAbstractDataSource.createPhysicalConnection(DruidAbstractDataSource.java:1709)
	at com.alibaba.druid.pool.DruidDataSource$CreateConnectionThread.run(DruidDataSource.java:2715)
Caused by: java.net.SocketException: Broken pipe (Write failed)
	at java.net.SocketOutputStream.socketWrite0(Native Method)
	at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:111)
	at java.net.SocketOutputStream.write(SocketOutputStream.java:155)
	at com.github.housepower.jdbc.buffer.SocketBuffedWriter.flushToTarget(SocketBuffedWriter.java:53)
	at com.github.housepower.jdbc.serializer.BinarySerializer.flushToTarget(BinarySerializer.java:79)
	at com.github.housepower.jdbc.connect.PhysicalConnection.disPhysicalConnection(PhysicalConnection.java:99)
```
### Other descriptions
resolve method:
class: com.github.housepower.jdbc.connect.PhysicalConnection
```
public void disPhysicalConnection(boolean isPingFail) throws SQLException {
        try {
            if (!socket.isClosed()) {
                if (isPingFail) {
                    serializer.flushToTarget(true);
                } else {
                    serializer.bufPositionTo0();
                }
                socket.close();
            }
        } catch (IOException ex) {
            throw new SQLException(ex.getMessage(), ex);
        }
    }
```
add method:
classes:
1. com.github.housepower.jdbc.serializer.BinarySerializer
2. com/github/housepower/jdbc/buffer/BuffedWriter#bufPositionTo0
BuffedWriter is a interface
3. com.github.housepower.jdbc.buffer.CompressedBuffedWriter#bufPositionTo0
4. com.github.housepower.jdbc.buffer.SocketBuffedWriter#bufPositionTo0
class CompressedBuffedWriter and SocketBuffedWriter is the completes implements BuffedWriter 
```
    public void bufPositionTo0() {
        container.get().bufPositionTo0();
    }
```

**Comment by jiakai908** [comment]
i also meet the same problem. how can i solve this.

**Comment by sundy-li** [comment]
@jiakai908  try the latest version.

**Comment by pan3793** [comment]
@sundy-li Have a quick check on mainline code, this issue should still exist.
@wangshuzheng I think I get your point, correct me if I miss understood. 

`socket.isClosed()` return a state which maintained by our side, if the socket broken, in other word, the pipe is cut off but we doesn't receive a FIN packet, `socket.isClosed()` will return `true` since the tcp connection status still is ESTABLISHED, then if we send packet, will get `java.net.SocketException: Broken pipe (Write failed)`.

If things just like I said above, based on your provided code, close and drop the data in buffer silently maybe not a ideal solution. I think we must let user know what happened, and let user to handle the data in buffer. 

But in some case, i.e. not data in buffer, I think close the broken socket silently is safe.

BTW, @wangshuzheng @jiakai908 would you please provide more information about your scenario? I want know what cause the socket broken?

**Comment by pan3793** [comment]
Maybe relative issue https://github.com/ClickHouse/clickhouse-jdbc/pull/515

**Comment by pan3793** [comment]
If should be fixed in v2.4.2, please have a try.
